### PR TITLE
[FIRRTL] Track clock values of memory ports precisely

### PR DESCRIPTION
**Pull request by uenoku**
We should identify the clocks to be same even if the clock is passed via different wires, This should prevent creating multiple write ports for combmem. We should consider to revert this change once we stop creating anoymous wires at FlattenMemory.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 1]
```diff
@@ -1922,6 +1922,24 @@ Value MemOp::getPortNamed(StringAttr name) {
   return Value();
 }
 
+static Value trackValueThroughAnoymousWires(Value value) {
+  auto wire = value.getDefiningOp<WireOp>();
+  // If the value is not a wire, return the value.
+  if (!wire)
+    return value;
+
+  // If the wire has a symbol, return the wire.
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1931]: It's not clear that symbols should matter here.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 2]
```diff
@@ -1922,6 +1922,24 @@ Value MemOp::getPortNamed(StringAttr name) {
   return Value();
 }
 
+static Value trackValueThroughAnoymousWires(Value value) {
+  auto wire = value.getDefiningOp<WireOp>();
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1926]: Nodes and wires are basically interchangable, we should track through them too I think

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 3]
```diff
@@ -1922,6 +1922,24 @@ Value MemOp::getPortNamed(StringAttr name) {
...
+  // If the value is not a wire, return the value.
+  if (!wire)
+    return value;
+
+  // If the wire has a symbol, return the wire.
+  if (wire.inner_sym() && !wire.inner_sym().getValue().empty())
+    return value;
+
+  for (auto *c : value.getUsers())
+    if (auto connect = dyn_cast<FConnectLike>(c))
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1936]: Are we out of last-connect semantics by this point?  If not, there might be multiple connects and this only checks the first one.

Comment (darthscsi) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1936]: If nothing else, we should be checking that there is only one connect.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 1 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: Thanks, I changed to use `getModuleScopeDriver` so it is fixed

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 2 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: Node is also tracked via `getModuleScopeDriver`.

**Comment by uenoku** [comment]
Considering the comment https://github.com/llvm/circt/pull/1805#discussion_r712690347, this PR might be controversial.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 3 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: `getModuleScopeDriver` assumes sigle connection so I'll follow-up to add assertions there.

**Comment by darthscsi** [comment]
I would say this is clearly correct (famous last words).

**Review by youngar — APPROVED** [review]
LGTM 👍

**Comment by uenoku** [comment]
Thanks, then I'd like to merge this.

**Pull request merged**